### PR TITLE
Node.js ES6 module support

### PR DESCRIPTION
**Pull request by nzakas**

### CODE REVIEW

File: manuscript/13-Modules.md [hunk 1]
```diff
@@ -456,13 +456,186 @@ This example loads `module.js` as a module instead of a script by passing a seco
...
 
-<!--
-    Note: Node.js is still debating the right way to load modules. I'd really like to include that description in this chapter and it looks like they will be getting close to resolution soon (in the next few weeks). I'd like to leave a placeholder here to come back after initial edits are complete to write this section.
-
 ### Using Modules in Node.js
 
-TODO
--->
+While adding support for modules in web browsers was fairly straightforward, adding modules to Node.js was a bit more involved. From the beginning, Node.js has supported the CommonJS module system and would need to continue supporting that format for the foreseeable future. Unlike in the browser, where JavaScript modules would only ever import other JavaScript modules, Node.js needed to support JavaScript modules importing CommonJS modules and CommonJS modules importing JavaScript modules. The end result was to create a new file extension, `.mjs`, to use instead of `.js` whenever a file contains an JavaScript module.
```
Comment (bmeck) [inline manuscript/13-Modules.md]: note: whatwg.github.io/loader/ is intended to support non ES modules in the browser.

File: manuscript/13-Modules.md [hunk 2]
```diff
@@ -456,13 +456,186 @@ This example loads `module.js` as a module instead of a script by passing a seco
...
+A>
+A> ```js
+A> import { sum } from "example/sum.js";
+A> ```
+A>
+A> While this code may look simple to understand, browsers and Node.js interpret `"example/sum.js"` differently. The differences are outlined in the table below.
+A>
+A> | **Environment** | **File to Load** | **Loaded As** | **Mode** |
+A> -----------------------------------------------------------------
+A> | Browser         | `./example/sum.js` | JavaScript module | strict |
```
Comment (bmeck) [inline manuscript/13-Modules.md]: Bare paths ala "example/" are banned by https://html.spec.whatwg.org/multipage/webappapis.html#resolve-a-module-specifier , but http://whatwg.github.io/loader/ will provide hooks to handle them.

So this does not load `./example/sum.js`, it errors out.

File: manuscript/13-Modules.md [hunk 3]
```diff
@@ -456,13 +456,186 @@ This example loads `module.js` as a module instead of a script by passing a seco
...
+A> | Node.js         | `./node_modules/example/sum.js` | CommonJS module | nonstrict |
+A>
+A> In a browser, `"example"` is resolved as a subdirectory whereas Node.js resolves `"example"` as a package name. Further, because the `.js` file extension is specified (which is required by the browser), the browser loads the file as a JavaScript module while Node.js loads the file as CommonJS. These differences pose some interoperability concerns for those who want to write JavaScript modules that work in both the browser and Node.js. The upcoming module loader specification is intended to help solve some of these module resolution interoperability concerns.
```
Comment (bmeck) [inline manuscript/13-Modules.md]: Note: browsers are not determining the type of file by extension, they only use the MIME in the content-type header, this MIME is shared for Script and Module goal of the language; it just assumes Module. This assumption has a grammatical ambiguity which means things meant for Script run in the wrong mode ([and parse/run, but produce incorrect values](https://gist.github.com/bmeck/e54d607a4b0874e8dfca7af2a132b9c2))...

Its pretty easy to see this in action by loading something like `<script src=my.php>` with any JS MIME in https://html.spec.whatwg.org/multipage/scripting.html#scriptingLanguages

File: manuscript/13-Modules.md [hunk 4]
```diff
@@ -456,13 +456,186 @@ This example loads `module.js` as a module instead of a script by passing a seco
...
+
+#### Loading CommonJS Modules from JavaScript Modules
+
+Since there are millions of existing Node.js modules written in CommonJS format, the ability for JavaScript modules to load CommonJS modules is important. As previously mentioned, an `import` statement is capable of loading from a CommonJS module, although the semantics are a bit different due to how CommonJS defines modules. Suppose that the following CommonJS module is in the file `example.js`:
```
Comment (bmeck) [inline manuscript/13-Modules.md]: hundreds of thousands, not quite millions :)

File: manuscript/13-Modules.md [hunk 5]
```diff
@@ -456,13 +456,186 @@ This example loads `module.js` as a module instead of a script by passing a seco
...
+
+The `sum()` function is imported directly from `example.js` as if the imported module were a JavaScript module. Similarly, you can import everything defined on `module.exports` by importing `*`, such as:
+
+```js
+import * as example from "./example";
+
+let result = example.sum(1, 2);
+```
+
+Here, the local binding `example` is a new object that has references to every own property on `module.exports` object from `example.js` (note that `example` is not the same object as `module.exports`). That means you can access the `sum()` function as a method on `example`.
```
Comment (bmeck) [inline manuscript/13-Modules.md]: new is misleading, the object is shared between imports,

if `a` imports `example` and `b` imports `example`, they would share the same namespace object.

File: manuscript/13-Modules.md (hunk 2 above)
Comment (nzakas) [inline manuscript/13-Modules.md]: Oh cool, totally missed that. Thanks for the references.

File: manuscript/13-Modules.md (hunk 4 above)
Comment (TehShrike) [inline manuscript/13-Modules.md]: hundreds of thousands on npm - millions in private repos, I'd expect!

File: manuscript/13-Modules.md (hunk 5 above)
Comment (nzakas) [inline manuscript/13-Modules.md]: Good point, thanks.

File: manuscript/13-Modules.md (hunk 3 above)
Comment (nzakas) [inline manuscript/13-Modules.md]: Right, I just worded the sentence wrong. I was trying to make the point that Node.js would act differently but got the sentence turned around. I'll fix that.

**Comment by bmeck** [comment]
lgtm

**Comment by nzakas** [comment]
Thanks @bmeck

**Pull request merged**